src/main/java/org/truffleruby/language/FrameSendingNode.java

```diff
@@ -73,6 +73,9 @@ protected synchronized void resetNeedsCallerAssumption() {
     public MaterializedFrame getFrameIfRequired(VirtualFrame frame) {
         switch (sendsFrame) {
             case MY_FRAME:
+                if (frame == null) {
```

Comment:
I do not like these changes. I had to do this to prove the main idea, but let's think about how we can avoid this.

I know the reason for this:
`LookupSuperMethodNode # getDeclarationContext` executes` readCallerFrame.execute (frame) `, which is probably not pure. I do not understand, but for some reason, if we execute this code, later `getFrameIfRequired` will be executed with an empty frame.  
This happens not in the test code itself, but earlier, when mspec is initialized. If we run just empty `jt ruby test.rb`, there is not error. Maybe you have an idea why this is so happening?

Reply:
ReadCallerFrameNode#execute() should not have visible side effects (it will just notify the caller to always send it the frame, so it doesn't have to walk the stack the next time it's used).
How can I reproduce that case?

Reply:
Just comment this `if` statement, simply do

```sh
jt build
jt test fast
```

and you will see the error (not from the LookupSuperMethodNode).

Moreover, I tried change `LookupSuperMethodNode#getDeclarationContext` to
```java

private DeclarationContext getDeclarationContext(VirtualFrame frame, DynamicObject selfMetaClass) {
  final DeclarationContext context = RubyArguments.getDeclarationContext(frame);

  readCallerFrame.execute(frame); // just call it. Dont use the result

  return context;
}
```

and it has the same error;

Reply:
Not relevant more